packages/expo-updates/ios/EXUpdates/AppLoader/EXUpdatesFileDownloader.m

```diff
@@ -84,50 +84,49 @@ - (void)downloadManifestFromURL:(NSURL *)url
                                                      timeoutInterval:EXUpdatesDefaultTimeoutInterval];
   [self _setManifestHTTPHeaderFields:request];
   [self _downloadDataWithRequest:request successBlock:^(NSData *data, NSURLResponse *response) {
+      
+    NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse *)response;
+    NSDictionary* headerDictionary = [httpResponse allHeaderFields];
+    NSString* headerSignature = [headerDictionary valueForKey:@"expo-manifest-signature"];
+      
     NSError *err;
     id parsedJson = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&err];
     if (err) {
       errorBlock(err, response);
       return;
     }
 
-    NSDictionary *manifest = [self _extractManifest:parsedJson error:&err];
+    NSDictionary* updateResponseDictionary = [self _extractUpdateResponseDictionary:parsedJson error:&err];
     if (err) {
       errorBlock(err, response);
       return;
     }
 
-    id innerManifestString = manifest[@"manifestString"];
-    id signature = manifest[@"signature"];
-    BOOL isSigned = innerManifestString != nil && signature != nil;
+    id bodyManifestString = updateResponseDictionary[@"manifestString"];
+    id bodySignature = updateResponseDictionary[@"signature"];
+    BOOL isSignatureInBody = bodyManifestString != nil && bodySignature != nil;
 
+    NSString* signature = isSignatureInBody ? bodySignature : headerSignature;
+    NSString* manifestString = isSignatureInBody ? bodyManifestString : [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+      
     // XDL serves unsigned manifests with the `signature` key set to "UNSIGNED".
     // We should treat these manifests as unsigned rather than signed with an invalid signature.
-    if (isSigned && [signature isKindOfClass:[NSString class]] && [(NSString *)signature isEqualToString:@"UNSIGNED"]) {
-      isSigned = NO;
-
-      NSError *err;
-      manifest = [NSJSONSerialization JSONObjectWithData:[(NSString *)innerManifestString dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&err];
-      NSAssert(!err && manifest && [manifest isKindOfClass:[NSDictionary class]], @"manifest should be a valid JSON object");
-      NSMutableDictionary *mutableManifest = [manifest mutableCopy];
-      mutableManifest[@"isVerified"] = @(NO);
-      manifest = [mutableManifest copy];
-    }
-
-    if (isSigned) {
-      NSAssert([innerManifestString isKindOfClass:[NSString class]], @"manifestString should be a string");
+    BOOL isUnsignedFromXDL = [(NSString *)signature isEqualToString:@"UNSIGNED"];
+
+    NSDictionary* manifest = [NSJSONSerialization JSONObjectWithData:[(NSString*)manifestString dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&err];
```

Comment:
let's add `isKindOfClass:[NSString class]` typechecks before casting these variables to NSString